{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Bring in mods and write ftns for analysis\n",
    "import pandas as pd, numpy as np, os, glob\n",
    "from Bio import SeqIO\n",
    "\n",
    "def dotfill(r,a,fill='.'):\n",
    "    return ''.join([fill for i in range(abs(len(a)-len(r)))])\n",
    "\n",
    "def scorealign(r,a,fill='.',verbose=False):\n",
    "    score = 0\n",
    "    ns = []\n",
    "    for i in range(len(r))[1:-1]:\n",
    "        temp = r[:i] + dotfill(r,a) + r[i:]\n",
    "        c = 0\n",
    "        for j,n in enumerate(a):\n",
    "            if n == temp[j]:\n",
    "                c = c + 1\n",
    "        if c > score:\n",
    "            ns = temp\n",
    "            score = c\n",
    "            if verbose:\n",
    "                print(ns,c)\n",
    "    if len(ns) == 0:\n",
    "        #print('Error empty sequence\\n%s\\n%s'%(r,a))\n",
    "        ns = r[0]+dotfill(r,a)+r[-1]\n",
    "    return ns\n",
    "\n",
    "def calcpos(seq,start,end,pad=0,fill='.'):\n",
    "    c = 0;\n",
    "    pos = np.arange(start,end+pad)\n",
    "    poss= []\n",
    "    if fill in seq:\n",
    "        for n in seq:\n",
    "            if n != fill:\n",
    "                poss.append(pos[c])\n",
    "                c = c + 1\n",
    "            else:\n",
    "                poss.append(pos[c])\n",
    "    else:\n",
    "        poss = pos\n",
    "    return poss\n",
    "\n",
    "def realign(ref,alt,verbose=False):\n",
    "    if len(alt) > len(ref):\n",
    "        newref = scorealign(ref,alt,verbose=verbose)\n",
    "        newalt = alt\n",
    "    elif len(ref) > len(alt):\n",
    "        newalt = scorealign(alt,ref,verbose=verbose)\n",
    "        newref = ref\n",
    "    else:\n",
    "        assert len(alt) == len(ref)\n",
    "        newref = ref\n",
    "        newalt = alt\n",
    "    return newref,newalt\n",
    "\n",
    "rdict = dict(zip(['three_prime_UTR',\n",
    "                  'five_prime_UTR',\n",
    "                  'CDS'],\n",
    "                 [3,5,0]))\n",
    "\n",
    "def gvtodf(gv,refseqdf,gffdf,rs='REF',fill='.',rdict=rdict,verbose=False):\n",
    "\n",
    "    refs = []\n",
    "    alts = []\n",
    "    poss = []\n",
    "    for i,j in gv.iterrows():\n",
    "    \n",
    "        ref = j.Ref\n",
    "        alt = j.Alt.split('.')[int(j.Sample)]\n",
    "        start = j.Pos\n",
    "        end = start + len(ref)\n",
    "    \n",
    "        newref,newalt = realign(ref,alt)\n",
    "        newpos = calcpos(newref,start,end)\n",
    "    \n",
    "        refs.append(newref)\n",
    "        alts.append(newalt)\n",
    "        poss.append(newpos)\n",
    "    \n",
    "    pos = np.concatenate(poss)\n",
    "    ref = [n for n in ''.join(refs)]\n",
    "    alt = [n for n in ''.join(alts)]\n",
    "\n",
    "    df = pd.DataFrame([pos,ref,alt],\n",
    "                      index=['Pos','Ref','Alt']).T\n",
    "    df['Isvar'] = 1\n",
    "\n",
    "    check = refseqdf[(refseqdf.Pos.isin(df.Pos))]\n",
    "    failed = []\n",
    "    for i,j in check.iterrows():\n",
    "        \n",
    "        tocheck = [s for s in \n",
    "                   df[(df.Pos==j.Pos)].Ref.tolist() \n",
    "                   if (s != fill)]\n",
    "        if len(tocheck) > 1: ## Should be one base\n",
    "            failed.append(j.Pos)\n",
    "        elif not tocheck[0] == j[rs]:\n",
    "            failed.append(j.Pos)\n",
    "            \n",
    "    if (len(failed) > 0) & verbose:\n",
    "        print('Reference alignment failed')\n",
    "    \n",
    "    dfadd = refseqdf[~(refseqdf.Pos.isin(df.Pos))].copy()\n",
    "    dfadd['Ref'] = dfadd[rs]\n",
    "    dfadd['Alt'] = dfadd[rs]\n",
    "    dfadd['Isvar'] = 0\n",
    "    dfadd.drop(rs,axis=1,inplace=True)\n",
    "\n",
    "    res = pd.concat([dfadd,df]).sort_values('Pos')\n",
    "    #res['Strand'] = -1 if gffdf.Strand.min() == '-' else 1\n",
    "    res['Strand'] = gffdf.Strand.min()\n",
    "    res['Type'] = -1\n",
    "    res['Phase'] = -1\n",
    "\n",
    "    for i,j in gffdf[(gffdf.Type!='gene')].iterrows():\n",
    "    \n",
    "        nb = np.arange(j.Start,j.End+1)\n",
    "        res.loc[(res.Pos.isin(nb)),'Type'] = j.Type\n",
    "    \n",
    "        if j.Type == 'CDS':\n",
    "            res.loc[(res.Pos.isin(nb)),'Phase'] = int(j.Phase)\n",
    "    \n",
    "    assert -1 not in res[(res.Type=='CDS')].Phase.tolist()\n",
    "    \n",
    "    res.Type.replace(rdict,inplace=True)\n",
    "\n",
    "    return res,failed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Seqid</th>\n",
       "      <th>Length</th>\n",
       "      <th>Cumpos</th>\n",
       "      <th>Midpts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>CP003820.1</td>\n",
       "      <td>2291499</td>\n",
       "      <td>0</td>\n",
       "      <td>1145749.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>CP003821.1</td>\n",
       "      <td>1621675</td>\n",
       "      <td>2291499</td>\n",
       "      <td>3102336.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>CP003822.1</td>\n",
       "      <td>1575141</td>\n",
       "      <td>3913174</td>\n",
       "      <td>4700744.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>CP003823.1</td>\n",
       "      <td>1084805</td>\n",
       "      <td>5488315</td>\n",
       "      <td>6030717.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>CP003824.1</td>\n",
       "      <td>1814975</td>\n",
       "      <td>6573120</td>\n",
       "      <td>7480607.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chrom       Seqid   Length   Cumpos     Midpts\n",
       "0      1  CP003820.1  2291499        0  1145749.5\n",
       "1      2  CP003821.1  1621675  2291499  3102336.5\n",
       "2      3  CP003822.1  1575141  3913174  4700744.5\n",
       "3      4  CP003823.1  1084805  5488315  6030717.5\n",
       "4      5  CP003824.1  1814975  6573120  7480607.5"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chrommap_path = '../../GENOTYPE/H99_chrommap.csv'\n",
    "chrommap = pd.read_csv(chrommap_path)\n",
    "chrommap.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Seqid</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Qual</th>\n",
       "      <th>Callrate</th>\n",
       "      <th>Nallele</th>\n",
       "      <th>Alleles</th>\n",
       "      <th>Maxlen</th>\n",
       "      <th>Minlen</th>\n",
       "      <th>Depth</th>\n",
       "      <th>Type</th>\n",
       "      <th>...</th>\n",
       "      <th>PMY2567</th>\n",
       "      <th>PMY2868</th>\n",
       "      <th>PMY2659</th>\n",
       "      <th>PMY2879</th>\n",
       "      <th>PMY2882</th>\n",
       "      <th>PMY2887</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>MAF</th>\n",
       "      <th>Mdepth</th>\n",
       "      <th>MAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP003824.1</td>\n",
       "      <td>1015010</td>\n",
       "      <td>332131.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>AAGTC.AAGTT</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>10745</td>\n",
       "      <td>snp</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>34.413681</td>\n",
       "      <td>0.951384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP003824.1</td>\n",
       "      <td>1015035</td>\n",
       "      <td>333883.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>C.T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>10643</td>\n",
       "      <td>snp</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>34.508143</td>\n",
       "      <td>0.955241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP003824.1</td>\n",
       "      <td>1015148</td>\n",
       "      <td>295782.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>TAACA.TAGCA</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>9297</td>\n",
       "      <td>snp</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>29.811075</td>\n",
       "      <td>0.954066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP003824.1</td>\n",
       "      <td>1015178</td>\n",
       "      <td>310092.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>CGAGGAAC.CGAAGAAC</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>9852</td>\n",
       "      <td>snp</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>31.485342</td>\n",
       "      <td>0.956476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP003824.1</td>\n",
       "      <td>1015276</td>\n",
       "      <td>241338.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGAGAAGG.AGGAAAAGG</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>8448</td>\n",
       "      <td>snp</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.869707</td>\n",
       "      <td>0.947224</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 321 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Seqid      Pos      Qual  Callrate  Nallele              Alleles  \\\n",
       "0  CP003824.1  1015010  332131.0       1.0        1          AAGTC.AAGTT   \n",
       "1  CP003824.1  1015035  333883.0       1.0        1                  C.T   \n",
       "2  CP003824.1  1015148  295782.0       1.0        1          TAACA.TAGCA   \n",
       "3  CP003824.1  1015178  310092.0       1.0        1    CGAGGAAC.CGAAGAAC   \n",
       "4  CP003824.1  1015276  241338.0       1.0        1  AGGAGAAGG.AGGAAAAGG   \n",
       "\n",
       "   Maxlen  Minlen  Depth Type  ...  PMY2567  PMY2868  PMY2659  PMY2879  \\\n",
       "0       5       5  10745  snp  ...      1.0      1.0      1.0      1.0   \n",
       "1       1       1  10643  snp  ...      1.0      1.0      1.0      1.0   \n",
       "2       5       5   9297  snp  ...      1.0      1.0      1.0      1.0   \n",
       "3       8       8   9852  snp  ...      1.0      1.0      1.0      1.0   \n",
       "4       9       9   8448  snp  ...      1.0      1.0      1.0      1.0   \n",
       "\n",
       "   PMY2882  PMY2887  Chrom  MAF     Mdepth       MAR  \n",
       "0        1        1      5  1.0  34.413681  0.951384  \n",
       "1        1        1      5  1.0  34.508143  0.955241  \n",
       "2        1        1      5  1.0  29.811075  0.954066  \n",
       "3        1        1      5  1.0  31.485342  0.956476  \n",
       "4        1        1      5  1.0  25.869707  0.947224  \n",
       "\n",
       "[5 rows x 321 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#geno_path = '../../GENOTYPE/Bt22xFtc555-1_genotypes.csv.gz'\n",
    "geno_path = '../../GENOTYPE/Bt22xFtc555-1_QTL_genotypes.csv.gz'\n",
    "genos = pd.read_csv(geno_path,index_col=0)\n",
    "genos = genos.merge(chrommap[['Chrom','Seqid']])\n",
    "\n",
    "genos.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Seqid</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Qual</th>\n",
       "      <th>Callrate</th>\n",
       "      <th>Nallele</th>\n",
       "      <th>Alleles</th>\n",
       "      <th>Maxlen</th>\n",
       "      <th>Minlen</th>\n",
       "      <th>Depth</th>\n",
       "      <th>Type</th>\n",
       "      <th>...</th>\n",
       "      <th>PMY2567</th>\n",
       "      <th>PMY2868</th>\n",
       "      <th>PMY2659</th>\n",
       "      <th>PMY2879</th>\n",
       "      <th>PMY2882</th>\n",
       "      <th>PMY2887</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>MAF</th>\n",
       "      <th>Mdepth</th>\n",
       "      <th>MAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 321 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Seqid, Pos, Qual, Callrate, Nallele, Alleles, Maxlen, Minlen, Depth, Type, PMY2940, PMY2936, PMY2934, PMY2932, PMY2931, PMY2929, PMY2927, PMY2926, PMY2925, PMY2924, PMY2919, PMY2918, PMY2917, PMY2916, PMY2912, PMY2910, PMY2908, PMY2906, PMY2905, PMY2902, PMY2901, PMY2898, PMY2937, PMY2897, PMY2893, PMY2891, PMY2890, PMY2888, PMY2734, PMY2725, PMY2729, PMY2723, PMY2616, PMY2928, PMY2685, PMY2717, PMY2712, PMY2857, PMY2864, PMY2711, PMY2638, PMY2859, PMY2788, PMY2710, PMY2896, PMY2706, PMY2830, PMY2848, PMY2709, PMY2858, PMY2707, PMY2703, PMY2702, PMY2697, PMY2678, PMY2677, PMY2693, PMY2670, PMY2860, PMY2581, PMY2673, PMY2672, PMY2715, PMY2669, PMY2614, PMY2785, PMY2895, PMY2798, PMY2696, PMY2871, PMY2921, PMY2667, PMY2875, PMY2665, PMY2695, PMY2658, PMY2605, PMY2599, PMY2832, PMY2687, PMY2656, PMY2679, PMY2680, PMY2796, PMY2731, PMY2913, PMY2904, PMY2892, PMY2816, PMY2873, ...]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 321 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genos[(genos.Pos==761938)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not 'Ref' in genos.columns:\n",
    "    genos['Ref'] = [a.split('.')[0] for a in genos.Alleles]\n",
    "    \n",
    "if not 'Alt' in genos.columns:\n",
    "    genos['Alt'] = genos['Alleles']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Seqid</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Length</th>\n",
       "      <th>Cumpos</th>\n",
       "      <th>Midpts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP003824.1</td>\n",
       "      <td>EuPathDB</td>\n",
       "      <td>gene</td>\n",
       "      <td>409610</td>\n",
       "      <td>410982</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CNAG_01397;description=mitochondrial import...</td>\n",
       "      <td>CNAG_01397</td>\n",
       "      <td>CNAG_01397</td>\n",
       "      <td>5</td>\n",
       "      <td>1814975</td>\n",
       "      <td>6573120</td>\n",
       "      <td>7480607.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP003824.1</td>\n",
       "      <td>EuPathDB</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>409610</td>\n",
       "      <td>410982</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CNAG_01397-t26_1;Parent=CNAG_01397;descript...</td>\n",
       "      <td>CNAG_01397</td>\n",
       "      <td>CNAG_01397</td>\n",
       "      <td>5</td>\n",
       "      <td>1814975</td>\n",
       "      <td>6573120</td>\n",
       "      <td>7480607.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP003824.1</td>\n",
       "      <td>EuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>409610</td>\n",
       "      <td>410051</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon_CNAG_01397-E1;Parent=CNAG_01397-t26_1</td>\n",
       "      <td>CNAG_01397-t26_1</td>\n",
       "      <td>CNAG_01397</td>\n",
       "      <td>5</td>\n",
       "      <td>1814975</td>\n",
       "      <td>6573120</td>\n",
       "      <td>7480607.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP003824.1</td>\n",
       "      <td>EuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>410113</td>\n",
       "      <td>410259</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon_CNAG_01397-E2;Parent=CNAG_01397-t26_1</td>\n",
       "      <td>CNAG_01397-t26_1</td>\n",
       "      <td>CNAG_01397</td>\n",
       "      <td>5</td>\n",
       "      <td>1814975</td>\n",
       "      <td>6573120</td>\n",
       "      <td>7480607.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP003824.1</td>\n",
       "      <td>EuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>410314</td>\n",
       "      <td>410424</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon_CNAG_01397-E3;Parent=CNAG_01397-t26_1</td>\n",
       "      <td>CNAG_01397-t26_1</td>\n",
       "      <td>CNAG_01397</td>\n",
       "      <td>5</td>\n",
       "      <td>1814975</td>\n",
       "      <td>6573120</td>\n",
       "      <td>7480607.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Seqid    Source  Type   Start     End Score  Strand Phase  \\\n",
       "0  CP003824.1  EuPathDB  gene  409610  410982     .       1     .   \n",
       "1  CP003824.1  EuPathDB  mRNA  409610  410982     .       1     .   \n",
       "2  CP003824.1  EuPathDB  exon  409610  410051     .       1     .   \n",
       "3  CP003824.1  EuPathDB  exon  410113  410259     .       1     .   \n",
       "4  CP003824.1  EuPathDB  exon  410314  410424     .       1     .   \n",
       "\n",
       "                                           Attribute            Parent  \\\n",
       "0  ID=CNAG_01397;description=mitochondrial import...        CNAG_01397   \n",
       "1  ID=CNAG_01397-t26_1;Parent=CNAG_01397;descript...        CNAG_01397   \n",
       "2      ID=exon_CNAG_01397-E1;Parent=CNAG_01397-t26_1  CNAG_01397-t26_1   \n",
       "3      ID=exon_CNAG_01397-E2;Parent=CNAG_01397-t26_1  CNAG_01397-t26_1   \n",
       "4      ID=exon_CNAG_01397-E3;Parent=CNAG_01397-t26_1  CNAG_01397-t26_1   \n",
       "\n",
       "         Gene  Chrom   Length   Cumpos     Midpts  \n",
       "0  CNAG_01397      5  1814975  6573120  7480607.5  \n",
       "1  CNAG_01397      5  1814975  6573120  7480607.5  \n",
       "2  CNAG_01397      5  1814975  6573120  7480607.5  \n",
       "3  CNAG_01397      5  1814975  6573120  7480607.5  \n",
       "4  CNAG_01397      5  1814975  6573120  7480607.5  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Bring in GFF file\n",
    "gffpath = '../../REF/FungiDB-46_CneoformansH99.gff.gz'\n",
    "names = [\"Seqid\", \"Source\", \"Type\", \"Start\", \"End\", \"Score\", \n",
    "         \"Strand\", \"Phase\", \"Attribute\"]\n",
    "dtype = [\"str\",\"str\",\"str\",\"int\",\"int\",\"str\",\"str\",\"str\",\"str\"]\n",
    "\n",
    "gff = pd.read_csv(gffpath,comment='#',\n",
    "                  sep='\\t',header=None,\n",
    "                  names=names,dtype=dict(zip(names,dtype)))\n",
    "\n",
    "gff['Strand'] = gff['Strand'].replace(dict(zip(['-','+'],[-1,1])))\n",
    "gff['Parent'] = [a.split('Parent=')[-1].split(';')[0].split('ID=')[-1] \n",
    "                 for a in gff.Attribute ]\n",
    "\n",
    "gff['Gene'] = [a.split('-t26')[0] for a in gff.Parent]\n",
    "\n",
    "gff = gff.merge(chrommap)\n",
    "\n",
    "gff.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Seqid</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Length</th>\n",
       "      <th>Cumpos</th>\n",
       "      <th>Midpts</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54476</th>\n",
       "      <td>CP003820.1</td>\n",
       "      <td>EuPathDB</td>\n",
       "      <td>gene</td>\n",
       "      <td>100</td>\n",
       "      <td>5645</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CNAG_04548;description=hypothetical protein</td>\n",
       "      <td>CNAG_04548</td>\n",
       "      <td>CNAG_04548</td>\n",
       "      <td>1</td>\n",
       "      <td>2291499</td>\n",
       "      <td>0</td>\n",
       "      <td>1145749.5</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59322</th>\n",
       "      <td>CP003820.1</td>\n",
       "      <td>EuPathDB</td>\n",
       "      <td>gene</td>\n",
       "      <td>5928</td>\n",
       "      <td>7982</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CNAG_07303;description=hypothetical protein...</td>\n",
       "      <td>CNAG_07303</td>\n",
       "      <td>CNAG_07303</td>\n",
       "      <td>1</td>\n",
       "      <td>2291499</td>\n",
       "      <td>0</td>\n",
       "      <td>1145749.5</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51999</th>\n",
       "      <td>CP003820.1</td>\n",
       "      <td>EuPathDB</td>\n",
       "      <td>gene</td>\n",
       "      <td>8766</td>\n",
       "      <td>9603</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CNAG_07304;description=hypothetical protein</td>\n",
       "      <td>CNAG_07304</td>\n",
       "      <td>CNAG_07304</td>\n",
       "      <td>1</td>\n",
       "      <td>2291499</td>\n",
       "      <td>0</td>\n",
       "      <td>1145749.5</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56198</th>\n",
       "      <td>CP003820.1</td>\n",
       "      <td>EuPathDB</td>\n",
       "      <td>gene</td>\n",
       "      <td>10835</td>\n",
       "      <td>11664</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CNAG_00001;description=hypothetical protein</td>\n",
       "      <td>CNAG_00001</td>\n",
       "      <td>CNAG_00001</td>\n",
       "      <td>1</td>\n",
       "      <td>2291499</td>\n",
       "      <td>0</td>\n",
       "      <td>1145749.5</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63828</th>\n",
       "      <td>CP003820.1</td>\n",
       "      <td>EuPathDB</td>\n",
       "      <td>gene</td>\n",
       "      <td>11773</td>\n",
       "      <td>12393</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CNAG_07305;description=hypothetical protein</td>\n",
       "      <td>CNAG_07305</td>\n",
       "      <td>CNAG_07305</td>\n",
       "      <td>1</td>\n",
       "      <td>2291499</td>\n",
       "      <td>0</td>\n",
       "      <td>1145749.5</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Seqid    Source  Type  Start    End Score  Strand Phase  \\\n",
       "54476  CP003820.1  EuPathDB  gene    100   5645     .      -1     .   \n",
       "59322  CP003820.1  EuPathDB  gene   5928   7982     .      -1     .   \n",
       "51999  CP003820.1  EuPathDB  gene   8766   9603     .      -1     .   \n",
       "56198  CP003820.1  EuPathDB  gene  10835  11664     .      -1     .   \n",
       "63828  CP003820.1  EuPathDB  gene  11773  12393     .      -1     .   \n",
       "\n",
       "                                               Attribute      Parent  \\\n",
       "54476     ID=CNAG_04548;description=hypothetical protein  CNAG_04548   \n",
       "59322  ID=CNAG_07303;description=hypothetical protein...  CNAG_07303   \n",
       "51999     ID=CNAG_07304;description=hypothetical protein  CNAG_07304   \n",
       "56198     ID=CNAG_00001;description=hypothetical protein  CNAG_00001   \n",
       "63828     ID=CNAG_07305;description=hypothetical protein  CNAG_07305   \n",
       "\n",
       "             Gene  Chrom   Length  Cumpos     Midpts           Description  \n",
       "54476  CNAG_04548      1  2291499       0  1145749.5  hypothetical protein  \n",
       "59322  CNAG_07303      1  2291499       0  1145749.5  hypothetical protein  \n",
       "51999  CNAG_07304      1  2291499       0  1145749.5  hypothetical protein  \n",
       "56198  CNAG_00001      1  2291499       0  1145749.5  hypothetical protein  \n",
       "63828  CNAG_07305      1  2291499       0  1145749.5  hypothetical protein  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Gather genes\n",
    "genes = gff[(gff.Type=='gene')].sort_values(['Seqid','Start','Strand']).copy()\n",
    "\n",
    "genes['Description'] = [a.split('description=')[-1].split('%2C')[0] \n",
    "                        for a in genes.Attribute]\n",
    "\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Make list of descriptions and features\n",
    "descriptions = ['hypothetical protein',\n",
    "                'unspecified product',\n",
    "                'conserved hypothetical protein']\n",
    "\n",
    "foi = ['gene','three_prime_UTR','five_prime_UTR','CDS']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Bring in reference file\n",
    "refpath = '../../REF/FungiDB-46_CneoformansH99_Genome.fasta'\n",
    "\n",
    "REF = [s for s in SeqIO.parse(refpath,format='fasta')]\n",
    "len(REF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3950, 16)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Collect known genes\n",
    "known = genes[~(genes.Description.isin(descriptions))]\n",
    "known_genes = known.Parent.tolist()\n",
    "known.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4388, 16)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Collect unknown genes\n",
    "unknown = genes[(genes.Description.isin(descriptions))]\n",
    "unknown_genes = unknown.Parent.tolist()\n",
    "unknown.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert genes[~(genes.Parent.isin(unknown.Parent)) &\n",
    "      ~(genes.Parent.isin(known.Parent))].shape[0] == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "amoeba_qtl_8 = (8,680000,770000)\n",
    "amoeba_qtl_5 = (5,1015000,1110000)\n",
    "lag_37_5 = (5,1112000,1405000)\n",
    "\n",
    "\n",
    "qtl = [amoeba_qtl_8,amoeba_qtl_5,lag_37_5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "219"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qtl_genes = []\n",
    "\n",
    "for (chrom,left,right) in qtl:\n",
    "    qtlg = genes[(genes.Chrom==chrom) & \n",
    "                  (genes.Start<=right) & \n",
    "                  (genes.End>=left)\n",
    "                ].Parent.tolist()\n",
    "    \n",
    "    qtl_genes.append(qtlg)\n",
    "    \n",
    "qtlgenes = np.unique(np.concatenate(qtl_genes))\n",
    "len(qtlgenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>CNAG_01125</td>\n",
       "      <td>CNAG_01125-t26_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>CNAG_01082</td>\n",
       "      <td>CNAG_01082-t26_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471</th>\n",
       "      <td>CNAG_01118</td>\n",
       "      <td>CNAG_01118-t26_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>527</th>\n",
       "      <td>CNAG_01097</td>\n",
       "      <td>CNAG_01097-t26_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>650</th>\n",
       "      <td>CNAG_01170</td>\n",
       "      <td>CNAG_01170-t26_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene            Parent\n",
       "70   CNAG_01125  CNAG_01125-t26_1\n",
       "397  CNAG_01082  CNAG_01082-t26_1\n",
       "471  CNAG_01118  CNAG_01118-t26_1\n",
       "527  CNAG_01097  CNAG_01097-t26_1\n",
       "650  CNAG_01170  CNAG_01170-t26_1"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Gather parents\n",
    "parents = gff[(gff.Type.isin(foi)) & \n",
    "              (gff.Type!='gene') & \n",
    "              (gff.Gene.isin(qtlgenes))\n",
    "             ][['Gene','Parent']\n",
    "              ].drop_duplicates()\n",
    "parents.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "parental_pmy = ['PMY2649', 'PMY2650']\n",
    "\n",
    "samplename = parental_pmy[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/croth/Bt22xFtc555-1/GENOTYPE/GENES/PMY2650/'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "savepath = '/home/croth/Bt22xFtc555-1/GENOTYPE/GENES/'\n",
    "savepath_sample = savepath+'%s/'%samplename\n",
    "\n",
    "if not os.path.exists(savepath):\n",
    "    os.mkdir(savepath)\n",
    "\n",
    "if not os.path.exists(savepath_sample):\n",
    "    os.mkdir(savepath_sample)\n",
    "    \n",
    "savepath_sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Pos', 'Alleles', 'Type', 'Chrom', 'Seqid', 'Ref', 'Alt']"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "info_cols = [c for c in genos.columns if c[:3]!='PMY']\n",
    "info_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "202"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run_parents = parents.Parent.tolist()\n",
    "len(run_parents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Is the GFF zero based?\n",
    "zb = False\n",
    "\n",
    "## Do you want to print everything \n",
    "## to the screne\n",
    "## like a madman?!?!\n",
    "verbos = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "581"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## RUN ALL THE GENES WITH CDS\n",
    "chroms = [8]#[1, 2, 3, 5, 6, 10, 11]\n",
    "run_parents = gff[(gff.Type=='CDS') & (gff.Chrom.isin(chroms))].Parent.unique()\n",
    "len(run_parents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n",
      "Reference alignment failed\n"
     ]
    }
   ],
   "source": [
    "novars = []\n",
    "notype = []\n",
    "sofail = []\n",
    "for ssk1name in run_parents:\n",
    "#for ssk1name in known_genes:\n",
    "    genesave = savepath_sample+ssk1name+'.csv.gz'\n",
    "    \n",
    "    #if os.path.exists(genesave):\n",
    "    #    continue\n",
    "    ## Take gene info from gff file\n",
    "    ssk1df = gff[(gff.Type.isin(foi)) & \n",
    "                 (gff.Parent==ssk1name)\n",
    "                ].sort_values('Start').copy()\n",
    "    \n",
    "    ssk1df['Start'] = ssk1df['Start'] + (1 if zb else 0)\n",
    "    \n",
    "    assert len(ssk1df.Strand.unique())==1\n",
    "    if len(ssk1df.Type.unique())!=3:\n",
    "        notype.append(ssk1name+'\\n')\n",
    "        continue\n",
    "    ## For a sample of interest bring in data\n",
    "    ##sample = pd.read_csv(samplepath,index_col=0)\n",
    "    sample = genos.loc[genos[samplename].dropna().index,\n",
    "                       info_cols+[samplename]]\n",
    "    \n",
    "    sample['Sample'] = sample[samplename]\n",
    "    ## Locate genetic variants within a gene\n",
    "    gv = sample[(sample.Seqid==ssk1df.Seqid.min()) & \n",
    "            (sample.Pos>=ssk1df.Start.min()) & \n",
    "            (sample.Pos<=ssk1df.End.max())]\n",
    "    if gv.shape[0] == 0:\n",
    "        novars.append(ssk1name+'\\n')\n",
    "        #if verbos:\n",
    "        #    print('No variants within %s for strain %s\\n'%(\n",
    "        #    ssk1name,samplename))\n",
    "        continue\n",
    "    ## From the referecne take the sequence of interest\n",
    "    gene_seq = [s.seq[ssk1df.Start.min()-(0 if zb else 1):ssk1df.End.max()] \n",
    "            for s in REF if s.id == ssk1df.Seqid.min()][0]\n",
    "    ## Make a dataframe from these sequences\n",
    "    refseqdf = pd.DataFrame([np.arange(ssk1df.Start.min()+(1 if zb else 0),\n",
    "                                       ssk1df.End.max()+1),\n",
    "                     [n for n in str(gene_seq)]],\n",
    "                            index=['Pos','REF']).T;\n",
    "    ## Make gene variant dataframe\n",
    "    res,failed  = gvtodf(gv,refseqdf,ssk1df,verbose=verbos)\n",
    "    #if (len(failed)>0) and verbos:\n",
    "    #    print(ssk1name+'\\n')\n",
    "    #    sofail.append(ssk1name)\n",
    "    res['Gene'] = ssk1name\n",
    "    res['Sample'] = samplename\n",
    "    ## Save and print results\n",
    "    res.drop_duplicates().to_csv(genesave,index=False)\n",
    "    if verbos and (len(failed)>0):\n",
    "        sofail.append(genesave)\n",
    "        #sofail.append(ssk1name,res.Pos.unique().shape[0],res.shape[0],\n",
    "        #    res[(res.Isvar==1)].shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/croth/Bt22xFtc555-1/GENOTYPE/GENES/PMY2650/'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "savepath_sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sofail)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "117"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(novars)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "failed_path = savepath_sample+'FAILED8.csv'\n",
    "if (len(sofail) > 0):\n",
    "    open(failed_path,'w').writelines(sofail)\n",
    "    \n",
    "novars_path = savepath_sample+'NOVARS8.csv'\n",
    "if (len(novars) > 0):\n",
    "    open(novars_path,'w').writelines(novars)\n",
    "\n",
    "notype_path = savepath_sample+'NOTYPE8.csv'\n",
    "if (len(notype) > 0):\n",
    "    open(notype_path,'w').writelines(notype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/croth/Bt22xFtc555-1/GENOTYPE/GENES/PMY2650/NOVARS8.csv'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novars_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/croth/Bt22xFtc555-1/GENOTYPE/GENES/PMY2650/'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "savepath_sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "417"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(run_parents) - (len(novars) + len(notype))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
